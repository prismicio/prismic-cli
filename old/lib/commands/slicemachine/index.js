import {
  ls,
  develop,
  endpoint,
  setup,
  bootstrap,
  createSlice,
  help,
  addStorybook,
} from './commands';
import { ctx as context } from '../../context';

function index(ctx = context) {
  if (ctx.SliceMachine.setup) return setup();
  if (ctx.SliceMachine.bootstrap) return bootstrap();
  if (ctx.SliceMachine.develop) return develop();
  if (ctx.SliceMachine.ls) return ls();
  if (ctx.SliceMachine.createSlice) return createSlice();
  if (ctx.SliceMachine.addStorybook) return addStorybook();
  if (ctx.SliceMachine.apiEndpoint && typeof ctx.SliceMachine.apiEndpoint === 'boolean') return endpoint();
  return help();
}

export default index;
