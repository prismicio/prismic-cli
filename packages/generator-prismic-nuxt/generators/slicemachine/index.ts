import PrismicGenerator, {TemplateOptions} from '@prismicio/prismic-yeoman-generator'
import {Question} from 'yeoman-generator'
import modifyNuxtConfig from './modify-nuxt-config'
import * as chalk from 'chalk'

const {SM_FILE} = require('sm-commons/consts')

export default class PrismicNuxt extends PrismicGenerator {
  /**
   * initializing - Your initialization methods (checking current project state, getting configs, etc)
   * prompting - Where you prompt users for options (where youâ€™d call this.prompt())
   * configuring - Saving configurations and configure the project (creating .editorconfig files and other metadata files)
   * default - If the method name doesnâ€™t match a priority, it will be pushed to this group.
   * writing - Where you write the generator specific files (routes, controllers, etc)
   * conflicts - Where conflicts are handled (used internally)
   * install - Where installations are run (npm, bower)
   * end - Called last, cleanup, say good bye, etc
   */

  pm: 'npm' | 'yarn' | undefined

  constructor(argv: string | string[], opts: TemplateOptions) {
    super(argv, opts)

    if (this.destinationRoot().endsWith(this.path) === false) {
      this.destinationRoot(this.path)
    }

    if (opts.domain) {
      this.domain = opts.domain
    }
  }

  async prompting() {
    const base = new URL(this.prismic.base)

    if (!this.domain) {
      const validateRepositoryName = this.prismic.validateRepositoryName
      const domainPrompt: Question = {
        type: 'input',
        name: 'domain',
        message: 'Name your prismic repository: ',
        transformer(value: string) {
          const reponame = value ? chalk.cyan(value) : chalk.dim.cyan('repo-name')
          const msg = [
            chalk.dim(`${base.protocol}//`),
            reponame,
            chalk.dim(`.${base.hostname}`),
          ]
          return msg.join('')
        },
        validate: async (value: string) => {
          const result = await validateRepositoryName(value)
          return result === value || result
        },
      }
      this.domain = await this.prompt(domainPrompt).then(res => res.domain)
    }

    if (!this.pm) await this.promptForPackageManager()
  }

  async writing() {
    const pkgJson = {
      scripts: {
        slicemachine: 'start-slicemachine --port 9999',
      },
      dependencies: {
        '@nuxtjs/prismic': '^1.2.6',
        '@prismicio/vue': '^2.0.11',
        'nuxt-sm': '^0.0.6',
        'vue-essential-slices': '^0.3.0',
        'vue-slicezone': '^0.0.30',
      },
      devDependencies: {
        sass: '^1.35.1',
        'css-loader': '^5.2.6',
        'sass-loader': '^10.1.1',
        'slice-machine-ui': 'beta',
      },
    }

    this.fs.extendJSON(this.destinationPath('package.json'), pkgJson)

    const config = this.readDestination('nuxt.config.js')

    const updatedConfig = modifyNuxtConfig(config, this.domain)

    this.writeDestination('nuxt.config.js', updatedConfig)

    this.fs.copyTpl(this.templatePath(), this.destinationPath(), {
      domain: this.domain,
      defaultLibrary: 'vue-essential-slices',
    })

    // maybe rename sm file
    if (this.existsDestination('sm.json') && SM_FILE !== 'sm.json') {
      this.moveDestination('sm.json', SM_FILE)
    }
    const customTypes = this.readCustomTypesFrom('customtypes')

    return this.prismic.createRepository({
      domain: this.domain,
      framework: 'nuxt',
      customTypes,
    }).then(res => {
      const url = new URL(this.prismic.base)
      url.host = `${res.data || this.domain}.${url.host}`
      this.log(`A new repository has been created at: ${url.toString()}`)
      return res
    })
  }

  async install() {
    if (this.pm === 'yarn') {
      this.yarnInstall()
    } else {
      this.npmInstall()
    }
  }

  async end() {
    const sliceZoneReadmeUrl = 'https://prismic.io/docs/technologies/vue-slicezone-technical-reference'

    const quickStartDocsUrl = 'https://prismic.io/docs/technologies/nuxtjs'

    const addSlicesDocsUrl = 'https://prismic.io/docs/technologies/generate-model-component-nuxtjs'

    const url = new URL(this.prismic.base)
    if (this.domain) {
      url.hostname = `${this.domain}.${url.hostname}`
      url.pathname = 'documents'
    } else {
      url.pathname = 'dashboard'
    }
    const writingRoomUrl = url.toString()

    return this.log(`
Your project is now configured to use SliceMachine!
Follow these next steps to get going:
    
- Add the SliceZone, anywhere in your code
${sliceZoneReadmeUrl}
    
- Access your Prismic writing room here
${writingRoomUrl}
    
- To add your own slice, run this command
$> npx prismic-cli sm --create-slice
    
- Run slicemachine
$> npm run slicemachine
    
If you're not sure where you should start,
give SliceMachine documentation a try:
    
- Quick Start ðŸ‘‰ ${quickStartDocsUrl}
- Adding slices ðŸ‘‰ ${addSlicesDocsUrl}
`)
  }
}

