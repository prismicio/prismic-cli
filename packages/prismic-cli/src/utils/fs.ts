// proxies to prevent mocking of node.js functions.
/**
 * import { readFile } from ../utls
 * */

import * as fs from 'fs'

export const readFile = fs.promises.readFile

export const writeFile = fs.promises.writeFile

export const unlink = fs.promises.unlink

export const rmdir = fs.promises.rmdir

export const mkdir = fs.promises.mkdir

export const readFileSync = fs.readFileSync

export const writeFileSync = fs.writeFileSync

export const existsSync = fs.existsSync

export const createWriteStream = fs.createWriteStream

export {copy} from 'fs-extra'
