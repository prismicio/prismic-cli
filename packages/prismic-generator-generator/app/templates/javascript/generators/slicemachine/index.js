const PrismicGenerator = require('@prismicio/prismic-yeoman-generator').default
const {SM_FILE} = require('sm-commons/consts')

class SliceMachine extends PrismicGenerator {
  /**
   * initializing - Your initialization methods (checking current project state, getting configs, etc)
   * prompting - Where you prompt users for options (where you’d call this.prompt())
   * configuring - Saving configurations and configure the project (creating .editorconfig files and other metadata files)
   * default - If the method name doesn’t match a priority, it will be pushed to this group.
   * writing - Where you write the generator specific files (routes, controllers, etc)
   * conflicts - Where conflicts are handled (used internally)
   * install - Where installations are run (npm, bower)
   * end - Called last, cleanup, say good bye, etc
   */

  constructor(argv, opts) {
    super(argv, opts)
    this.pm = opts.pm
    this.domain = opts.domain
    if (this.destinationRoot().endsWith(this.path) === false) {
      this.destinationRoot(this.path)
    }
  }

  async prompting() {
    if (!this.pm) await this.promptForPackageManager()
  }

  async writing() {
    const pkgJson = {
      scripts: {
        slicemachine: 'start-slicemachine --port 9999',
      },
      dependencies: {},
      devDependencies: {
        'slice-machine-ui': 'beta',
      },
    }

    this.fs.extendJSON(this.destinationPath('package.json'), pkgJson)

    this.fs.copyTpl(this.templatePath(), this.destinationPath(), {
      defaultLibrary: '',
      domain: this.domain,
    }, undefined, {
      globOptions: {dot: true},
    })

    // maybe rename sm file
    if (this.existsDestination('sm.json') && SM_FILE !== 'sm.json') {
      this.moveDestination('sm.json', SM_FILE)
    }

    const customTypes = this.readCustomTypesFrom('custom_types')
    return this.maybeCreatePrismicRepository({domain: this.domain, framework: '<%= name %>', customTypes}, this.existingRepo)
  }

  async install() {
    if (this.pm === 'yarn') {
      this.yarnInstall()
    } else {
      this.npmInstall()
    }
  }

  async end() {
    const url = new URL(this.prismic.base)

    if (this.domain) {
      url.hostname = `${this.domain}.${url.hostname}`
      url.pathname = 'documents'
    } else {
      url.pathname = 'dashboard'
    }

    const writingRoomUrl = url.toString()

    this.log(`
Your project is now configured to use SliceMachine!
Follow these next steps to get going:
    
- Add the SliceZone, anywhere in your code https://github.com/prismicio/slice-machine
- Access your Prismic writing room here
${writingRoomUrl}
    
- To add your own slice, run this command
$> npx prismic-cli sm --create-slice
    
- Run slicemachine
$> npx prismic-cli sm --develop
`)
  }
}

module.exports = SliceMachine