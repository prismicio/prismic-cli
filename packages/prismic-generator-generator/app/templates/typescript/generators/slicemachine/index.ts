import PrismicGenerator, {TemplateOptions} from '@prismicio/prismic-yeoman-generator'
const {SM_FILE} = require('sm-commons/consts')

export default class SliceMachine extends PrismicGenerator {
  /**
   * initializing - Your initialization methods (checking current project state, getting configs, etc)
   * prompting - Where you prompt users for options (where you’d call this.prompt())
   * configuring - Saving configurations and configure the project (creating .editorconfig files and other metadata files)
   * default - If the method name doesn’t match a priority, it will be pushed to this group.
   * writing - Where you write the generator specific files (routes, controllers, etc)
   * conflicts - Where conflicts are handled (used internally)
   * install - Where installations are run (npm, bower)
   * end - Called last, cleanup, say good bye, etc
   */
  framework: string;

  pm: 'npm' | 'yarn' | undefined

  constructor(argv: string | string[], opts: TemplateOptions) {
    super(argv, opts)

    if (this.destinationRoot().endsWith(this.path) === false) {
      this.destinationRoot(this.path)
    }
  }

  async prompting() {
    if (!this.pm) await this.promptForPackageManager()
  }

  async writing() {
    const pkgJson = {
      scripts: {
        slicemachine: 'start-slicemachine --port 9999',
      },
      dependencies: {},
      devDependencies: {
        'slice-machine-ui': 'beta',
      },
    }

    this.fs.extendJSON(this.destinationPath('package.json'), pkgJson)

    this.fs.copyTpl(this.templatePath(), this.destinationPath(), {
      defaultLibrary: '',
      domain: this.domain,
    }, undefined, {
      globOptions: {dot: true},
    })

    // maybe rename sm file
    if (this.existsDestination('sm.json') && SM_FILE !== 'sm.json') {
      this.moveDestination('sm.json', SM_FILE)
    }

    const customTypes = this.readCustomTypesFrom('custom_types')

    return this.prismic.createRepository({
      domain: this.domain,
      customTypes,
      framework: '<%= name %>'
    }).then(res => {
      const url = new URL(this.prismic.base)
      url.host = `${res.data || this.domain}.${url.host}`
      this.log(`A new repository has been created at: ${url.toString()}`)
      return res
    })
  }

  async install() {
    if (this.pm === 'yarn') {
      this.yarnInstall()
    } else {
      this.npmInstall()
    }
  }

  async end() {
    const url = new URL(this.prismic.base)

    if (this.domain) {
      url.hostname = `${this.domain}.${url.hostname}`
      url.pathname = 'documents'
    } else {
      url.pathname = 'dashboard'
    }

    const writingRoomUrl = url.toString()

    this.log(`
Your project is now configured to use SliceMachine!
Follow these next steps to get going:
    
- Add the SliceZone, anywhere in your code https://github.com/prismicio/slice-machine
- Access your Prismic writing room here
${writingRoomUrl}
    
- To add your own slice, run this command
$> npx prismic-cli sm --create-slice
    
- Run slicemachine
$> npx prismic-cli sm --develop
`)
  }
}

