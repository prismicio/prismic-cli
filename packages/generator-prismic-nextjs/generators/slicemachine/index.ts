import PrismicGenerator, {TemplateOptions} from '@prismicio/prismic-yeoman-generator'
import type {Question} from 'yeoman-generator'
import * as chalk from 'chalk'
import {existsSync} from 'fs'

const {SM_FILE} = require('sm-commons/consts')

export default class SliceMachine extends PrismicGenerator {
  /**
   * initializing - Your initialization methods (checking current project state, getting configs, etc)
   * prompting - Where you prompt users for options (where youâ€™d call this.prompt())
   * configuring - Saving configurations and configure the project (creating .editorconfig files and other metadata files)
   * default - If the method name doesnâ€™t match a priority, it will be pushed to this group.
   * writing - Where you write the generator specific files (routes, controllers, etc)
   * conflicts - Where conflicts are handled (used internally)
   * install - Where installations are run (npm, bower)
   * end - Called last, cleanup, say good bye, etc
   */
  framework: string;

  pm: 'npm' | 'yarn' | undefined

  constructor(argv: string | string[], opts: TemplateOptions) {
    super(argv, opts)

    if (this.destinationRoot().endsWith(this.path) === false) {
      this.destinationRoot(this.path)
    }

    if (opts.framework) {
      this.config.set('framework', opts.framework)
      this.framework = opts.framework
    } else {
      this.framework = this.config.get('framework') || 'nextjs'
      this.config.set('framework', this.framework)
    }

    if (opts.domain) {
      this.domain = opts.domain
    }
  }

  async prompting() {
    const base = new URL(this.prismic.base)
    
    if (!this.domain) {
      const validateRepositoryName = this.prismic.validateRepositoryName
      const domainPrompt: Question = {
        type: 'input',
        name: 'domain',
        message: 'Name your prismic repository: ',
        transformer(value: string) {
          const reponame = value ? chalk.cyan(value) : chalk.dim.cyan('repo-name')
          const msg = [
            chalk.dim(`${base.protocol}//`),
            reponame,
            chalk.dim(`.${base.hostname}`),
          ]
          return msg.join('')
        },
        validate: async (value: string) => {
          const result = await validateRepositoryName(value)
          return result === value || result
        },
      }
      this.domain = await this.prompt(domainPrompt).then(res => res.domain)
    }

    if (!this.pm) await this.promptForPackageManager()
  }

  async writing() {
    const pkgJson = {
      scripts: {
        slicemachine: 'start-slicemachine --port 9999',
      },
      dependencies: {
        'prismic-javascript': '^3.0.2',
        'prismic-reactjs': '^1.3.3',
        'next-slicezone': '^0.0.14',
        'next-transpile-modules': '^6.4.0',
        'theme-ui': '^0.6.2',
        'essential-slices': '^1.0.3',
      },
      devDependencies: {
        '@babel/core': '^7.12.10',
        'slice-machine-ui': '^0.0.45',
      },
    }

    this.fs.extendJSON(this.destinationPath('package.json'), pkgJson)

    // theses files could be removed from this package but would have to come from create-next-app
    const useSrc = existsSync(this.destinationPath('src'))
    
    this.fs.copyTpl(this.templatePath(), this.destinationPath(), {
      smFile: SM_FILE,
      useSrc,
      defaultLibrary: 'essential-slices',
      domain: this.domain,
      latest: '0.0.43',
    }, undefined, {
      globOptions: {dot: true},
    })

    if (useSrc) {
      this.deleteDestination('pages')
      this.fs.copyTpl(this.templatePath('next/pages'), this.destinationPath('src/pages'), {smFile: SM_FILE, useSrc}, undefined, {globOptions: {dot: true}})
    }

    // maybe rename sm file
    if (this.existsDestination('sm.json') && SM_FILE !== 'sm.json') {
      this.moveDestination('sm.json', SM_FILE)
    }

    const customTypes = this.readCustomTypesFrom('custom_types')

    return this.prismic.createRepository({
      domain: this.domain,
      customTypes,
    }).then(res => {
      const url = new URL(this.prismic.base)
      url.host = `${res.data || this.domain}.${url.host}`
      this.log(`A new repository has been created at: ${url.toString()}`)
      return res
    })
  }

  async install() {
    if (this.pm === 'yarn') {
      this.yarnInstall()
    } else {
      this.npmInstall()
    }
  }

  async end() {
    const url = new URL(this.prismic.base)

    if (this.domain) {
      url.hostname = `${this.domain}.${url.hostname}`
      url.pathname = 'documents'
    } else {
      url.pathname = 'dashboard'
    }

    const writingRoomUrl = url.toString()

    this.log(`
Your project is now configured to use SliceMachine!
Follow these next steps to get going:
    
- Add the SliceZone, anywhere in your code https://github.com/prismicio/slice-machine/tree/master/packages/next-slicezone
    
- Access your Prismic writing room here
${writingRoomUrl}
    
- To add your own slice, run this command
$> npx prismic-cli sm --create-slice
    
- Run slicemachine
$> npx prismic-cli sm --develop
    
If you're not sure where you should start,
give SliceMachine documentation a try:
    
- Quick Start ðŸ‘‰ https://prismic.io/docs/technologies/quick-start-nextjs
- Adding slices ðŸ‘‰ https://prismic.io/docs/technologies/create-your-own-slices-components-nextjs
`)
  }
}

