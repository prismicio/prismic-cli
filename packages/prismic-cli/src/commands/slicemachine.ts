import {flags} from '@oclif/command'
import env, {meta, filterMetaFor} from '../prismic/yeoman-env'
import Command from '../prismic/base-command'
import * as path from 'path'
import {fs} from '../utils'
import {execSync} from 'child_process'
import {lookpath} from 'lookpath'
import * as inquirer from 'inquirer'
import {detect, PkgJson} from '../utils/framework'

const globby = require('fast-glob')

const {SM_FILE} = require('sm-commons/consts')

export default class Slicemachine extends Command {
  static description = 'Slice Machine commands'

  static aliases = ['sm']

  static flags = {
    help: flags.help({char: 'h'}),

    force: flags.boolean({char: 'f'}),

    setup: flags.boolean({
      description: 'Setup Slice Machine in an already existing project.',
      exclusive: ['create-slice', 'add-storybook', 'list', 'bootstrap', 'develop'],
      default: false,
    }),

    domain: flags.string({
      char: 'd',
      description: 'Prismic repo to create.',
      exclusive: ['add-storybook', 'create-slice', 'sliceName', 'library', 'list', 'develop'],
    }),

    'create-slice': flags.boolean({
      description: 'Add a Slice to a Slice Machine project.',
      exclusive: ['setup', 'add-storybook', 'list', 'bootstrap', 'develop'],
      default: false,
    }),

    sliceName: flags.string({
      description: 'Name of the Slice.',
      dependsOn: ['create-slice'],
    }),

    library: flags.string({
      description: 'Name of the Slice library.',
      dependsOn: ['create-slice'],
    }),

    'add-storybook': flags.boolean({
      description: 'Add Storybook to a Slice Machine project.',
      exclusive: ['setup', 'create-slice', 'list', 'bootstrap', 'develop'],
      default: false,
    }),

    framework: flags.string({
      description: 'framework to use, see list for options',
      exclusive: ['list'],
    }),

    list: flags.boolean({
      description: 'List local Slices.',
      exclusive: ['add-storybook', 'setup', 'create-slice', 'bootstrap', 'sliceName', 'domain', 'library', 'framework', 'folder', 'skip-install', 'develop', 'existing-repo'],
      default: false,
    }),

    folder: flags.string({
      description: 'Output directory.',
    }),

    'skip-install': flags.boolean({
      description: 'Prevent npm install from running.',
      exclusive: ['create-slice', 'list', 'bootstrap'],
      default: false,
    }),

    bootstrap: flags.boolean({
      description: 'Reconfigure a Slice Machine project.',
      exclusive: ['setup', 'create-slice', 'list', 'develop'],
      default: false,
    }),

    develop: flags.boolean({
      description: 'Run Slice Machine.',
      exclusive: ['setup', 'create-slice', 'list', 'bootstrap'],
      default: false,
    }),

    customTypeEndpoint: flags.string({
      description: 'Use a different custom-type endpoint.',
      hidden: true,
      dependsOn: ['develop'],
    }),

    'existing-repo': flags.boolean({
      description: 'Connect to an existing Prismic repository when running --setup or --bootstrap',
      default: false,
      exclusive: ['add-storybook', 'create-slice', 'develop'],
    }),
  }

  private async readGeneratorsFromToRc(folder: string): Promise<Array<string>> {
    const pathToYoRc = path.join(folder, '.yo-rc.json')
    const hasYoRc = fs.existsSync(pathToYoRc)

    if (!hasYoRc) return Promise.resolve([])

    const YoRc: Record<string, any> = JSON.parse(fs.readFileSync(pathToYoRc, 'utf-8'))
    const generatorNames = Object.keys(YoRc)
    .filter(d => d.startsWith('generator-prismic-'))
    .map(d => d.replace(/^generator-/, ''))

    return Promise.resolve(generatorNames)
  }

  async maybePromptForFrameWork(frameworksInYoRc: Array<string>, subGeneratorName: string): Promise<string> {
    const choices = (frameworksInYoRc.length === 0) ? (
      this.posibleFrameWorksForSubGeneratorAsPromtps(subGeneratorName)
    ) : frameworksInYoRc.map(d => ({
      name: d.replace(/^prismic-/, '').replace(/:.+/, ''),
      value: d.replace(/:.+/, ''),
    }))

    return inquirer.prompt<{framework: string}>({
      type: 'list',
      choices,
      name: 'framework',
      message: 'Which framework to use',
    }).then(res => res.framework)
  }

  private posibleFrameWorksForSubGeneratorAsPromtps(subGeneratorName: string) {
    const generatorsWithSubGenerators = filterMetaFor(meta, subGeneratorName)
    return Object.values(generatorsWithSubGenerators).map(d => {
      const name = d.namespace.replace(/^prismic-/, '').replace(/:.+/, '')
      const value = d.namespace.replace(/:.+/, '')
      return {name, value}
    })
  }

  private async runSubGenerator(name: string, folder: string, opts: Record<string, any>) {
    return this.readGeneratorsFromToRc(folder)
    .then(res => {
      if (res.length === 0 || res.length > 1) return this.maybePromptForFrameWork(res, name)
      return res[0]
    })
    .then(framework => {
      const setup = `${framework}:${name}`
      return this.envRun(setup, opts)
    })
  }

  private async handleSetup(folder: string, opts: Record<string, any>) {
    return this.runSubGenerator('slicemachine', folder, opts)
  }

  private async handleCreateSlice(folder: string, opts: Record<string, any>) {
    return this.runSubGenerator('create-slice', folder, opts)
  }

  private async handleStorybook(folder: string, opts: Record<string, any>) {
    return this.runSubGenerator('storybook', folder, opts)
  }

  async envRun(generatorName: string, options: Record<string, any>): Promise<void> {
    return new Promise((resolve, reject) => {
      return env.run(generatorName, options, err => {
        if (err) return reject(err)
        return resolve()
      })
    })
  }

  async run() {
    const {flags} = this.parse(Slicemachine)

    const folder = flags.folder || process.cwd()

    const opts = {
      ...flags,
      prismic: this.prismic,
      path: folder,
      existingRepo: flags['existing-repo'] || false,
    }

    if (flags['create-slice']) {
      return this.handleCreateSlice(folder, opts)
    }

    if (flags['add-storybook']) {
      return this.handleStorybook(folder, opts)
    }

    if (flags.setup) {
      const isAuthenticated = await this.prismic.isAuthenticated()
      if (!isAuthenticated) {
        await this.login()
      }

      const domain = await this.validateDomain(flags.domain, opts.existingRepo)

      return this.handleSetup(folder, {...opts, domain})
    }

    if (flags.list) {
      const pathToSMFile = path.join(folder, SM_FILE)
      if (fs.existsSync(pathToSMFile) === false) {
        this.error(`Could not find ${SM_FILE} at: ${pathToSMFile}`)
      }

      return fs.readFile(pathToSMFile, 'utf-8')
      .then(JSON.parse)
      .then(json => json.libraries || [])
      .then((libs: Array<string>) => {
        libs.forEach((lib: string) => {
          const isLocal = lib.startsWith('@/') || lib.startsWith('~') || lib.startsWith('/')
          const pathToSlices = path.posix.join(
            folder,
            isLocal ? '.' : 'node_modules',
            isLocal ? lib.substring(1, lib.length) : lib,
            '**',
            'model.json',
          )

          const pathToModels: Array<string> = globby.sync(pathToSlices)
          const names = pathToModels.map(pathToModel => {
            const dir = path.dirname(pathToModel)
            const dirs = dir.split(path.sep)
            return dirs[dirs.length - 1]
          })

          this.log(lib)
          names.forEach(name => this.log('\t' + name))
        })
      })
      .catch(this.error)
    }

    if (flags.bootstrap) {
      return this.handleBootStrap(folder, flags['existing-repo'], flags.domain)
    }

    if (flags.develop) {
      const isAuthenticated = await this.prismic.isAuthenticated()
      if (!isAuthenticated) await this.login()

      this.checkIsInASlicemachineProject()

      const hasYarn = await lookpath('yarn')
      const usingYarn = fs.existsSync(path.join(process.cwd(), 'yarn.lock'))
      const pm = hasYarn && usingYarn ? 'yarn' : 'npm run'
      return execSync(`${pm} slicemachine`, {stdio: 'inherit'})
    }

    if (!flags['create-slice'] && !flags['add-storybook'] && !flags.setup && !flags.list) {
      return this._help()
    }
  }

  private async handleBootStrap(folder: string, existingRepo = false, maybeDomain?: string) {
    const smFilePath = path.join(folder, SM_FILE)
    const pkgJsonPath = path.join(folder, 'package.json')
    const hasPackageJson = fs.existsSync(pkgJsonPath)
    const packageJson: PkgJson = hasPackageJson ? JSON.parse(fs.readFileSync(pkgJsonPath, 'utf-8')) : {}

    if (fs.existsSync(smFilePath) === false) {
      return this.warn(`${SM_FILE} file not found in: ${smFilePath}`)
    }

    const isAuthenticated = await this.prismic.isAuthenticated()
    if (!isAuthenticated) {
      await this.login()
    }

    const domain = await this.validateDomain(maybeDomain, existingRepo)
    const framework = await detect(packageJson) || ''

    const maybeCreateRepo = () => existingRepo ? Promise.resolve({data: domain}) : this.prismic.createRepository({domain, framework}) /* the framework is already registered on intercom, default value is '' for wroom */

    return maybeCreateRepo().then(res => {
      const url = new URL(this.prismic.base)
      url.hostname = `${res.data}.${url.hostname}`
      return this.log(`Your Slice Machine repository was successfully created! ${url.toString()}`)
    })
    .then(() => fs.readFile(smFilePath, 'utf-8'))
    .then(str => JSON.parse(str))
    .then(json => {
      const url = new URL(this.prismic.base)
      url.hostname = `${domain}.cdn.${url.hostname}`
      url.pathname = 'api/v2'
      return JSON.stringify({...json, apiEndpoint: url.toString()}, null, 2)
    }).then(smFile => {
      return fs.writeFile(smFilePath, smFile)
    })
  }

  private checkIsInASlicemachineProject(): string {
    const pathToSMFile = path.join(process.cwd(), SM_FILE)

    if (fs.existsSync(pathToSMFile) === false) {
      this.warn(`[slice-machine] Could not find ${pathToSMFile}`)
      return this.exit()
    }

    const smFile = fs.readFileSync(pathToSMFile, 'utf-8')
    try {
      const apiEndpoint: string = JSON.parse(smFile).apiEndpoint
      const url = new URL(apiEndpoint)
      return url.hostname.split('.')[0]
    } catch {
      this.warn(`[slice-machine] No "apiEndpoint" value found in ${pathToSMFile} .\nIn order to run this command, you need to set a Prismic repository endpoint`)
      return this.exit()
    }
  }
}
