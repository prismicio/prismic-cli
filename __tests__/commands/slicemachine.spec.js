/* eslint-disable global-require */
import fs from 'fs';
import mockConsole from 'jest-mock-console';
import path from 'path';
import globby from 'globby';
import consola from 'consola';
import slicemachine from '../../lib/commands/slicemachine';


describe('slice machine commands', () => {
  describe('help', () => {
    it('should be called by default', () => {
      const restoreConsole = mockConsole();

      jest.mock('../../lib/context');
      const context = require('../../lib/context');
      Object.assign(context, { SliceMachine: {} });

      slicemachine(context);

      expect(console.log).toHaveBeenCalled();

      restoreConsole();
      jest.unmock('../../lib/context');
    });
  });

  describe('ls', () => {
    const context = { SliceMachine: { ls: true } };

    it('should list libaries from...', async () => {
      jest.mock('fs');

      fs.existsSync = jest.fn();
      fs.existsSync.mockReturnValue(true);


      jest.mock('globby');
      globby.sync = jest.fn();
      globby.sync.mockReturnValue([
        path.resolve(__dirname, '../../node_modules/', 'vue-essential-slices/src/slices/CallToAction/model.json'),
      ]);

      const mockSMConfig = {
        // pathToSlices,
        // isLocal,
        libraries: [
          '@/slices',
          'vue-essential-slices',
        ],
      };

      fs.readFileSync = jest.fn();
      fs.readFileSync.mockReturnValue(JSON.stringify(mockSMConfig));

      jest.mock('console');
      console.error = jest.fn();
      console.log = jest.fn();

      consola.mockTypes(typeName => (typeName === 'info' && jest.fn()));

      const libs = await slicemachine(context);

      expect(fs.existsSync).toHaveBeenCalled();
      expect(fs.readFileSync).toHaveBeenCalled();
      expect(console.error).not.toHaveBeenCalled();
      expect(consola.info).toHaveBeenCalled();
      expect(console.log).toHaveBeenCalled(); // maybe match snapshot?

      expect(libs).toMatchSnapshot();


      jest.unmock('fs');
      jest.unmock('globby');
      jest.unmock('console');
    });
  });

  describe.skip('create slice', () => {
    it('should work', () => {
      jest.mock('../../common/frameworks');
      const Frameworks =  require('../../common/frameworks');
      Frameworks.get = jest.fn().mockReturnValue('nuxt');

      jest.mock('../../common/libraries');
      const Libraries = require('../../common/libraries');
      Libraries.selectLocal = jest.fn().mockResolvedValue('nuxt');
      Libraries.infos = jest.fn().mockReturnValue({
        config: '',
        isLocal: true,
        pathToLib: '',
        pathToSlices: '',
        relativePathToLib: '',
      });

      /* jest.mock('inquirer');
      const inquirer = require('inquirer');
      inquirer.prompt = jest.fn().mockResolvedValue('FooBar'); */

      jest.mock('fs');
      const fs = require('fs');
      fs.existsSync = jest.fn().mockReturnValue(true);
      fs.writeFileSync = jest.fn()

      jest.mock('console');
      console.error = jest.fn();
      console.log = jest.fn();
      
    });
  });
});

describe('bootstrap', () => {
  // boot-strap query string should have ?app=slicemachine 
  // it.skip('should work', () => {})

  it('should added app=slicemachine to the query string', () => {
    /*jest.mock('request');
    const request = require('request');
    request.post = jest.fn((url, options, callback) => (callback({}))*/
    // lib/services/communication/index.js
    jest.mock('../../lib/services/communication')
    const communication = require('../../lib/services/communication')
    communication.post = jest.fn().mockResolvedValue({ domain: 'foobar' });
  })
})


/** 'steup/init'
 * 
 * boot-strap query string should have ?app=slicemachine 
 * 
 * Frameworks get checks package.json for next or nuxt.
 * 
 * >downloads zip http://sm-api-p1g4hvtof.vercel.app/api/bootstrap?framework=nuxt
 * 
 * unzips and checks path to zip
 * 
 * exicutes the commands in either boot.json of bootstrap.json
 * 
 * boot.json should be the file to call
 * 
 * create a prismic repository with the custom types returned by the api.
 * 
 * an info object is returned by an internal function
 * 
 * installs dependancies
 * 
 * creates sm.json (a file that contains api endpoint, and an array of libraries that are used)
 * {
 *  "libraries": ["@/slices", "vue-essential-slices"],
 *  "apiEndpoint": "http://shared.wroom.test/api/v2",
 *  "dbId": "prismicio-94fce09d-e85f-4dc2-bf68-eacd2f5c2ad3_5"
 * }
 *
 * writes out the files from mainifest?
 * 
 ** handleRecap
 * > generates config
 * > transform config
 * > generates a display stating what needs to be done next
 * */

/** create-slice
  * creates a slice in the project.
  * 
  * > gets the framework
  * > reads sm.json looking for local libaries/directories or creates one for the slice.
  * > asks for slice name
  * > creates a remote slice
  * > creates a local slice
*/ 

/** Pull
  * gets custom-types
  * get the slice zones for custom types
  * gets all the model locally
  *
  * Diffs localfiles and remote files.
  * creates missing slices
  * prompts user about conflicting slices.
* */

/**
 * bootstrap boot-strap query string should have ?app=slicemachine 
*/
