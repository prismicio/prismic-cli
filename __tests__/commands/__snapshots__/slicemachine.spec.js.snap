// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slice machine commands ls should list libaries from... 1`] = `
Array [
  Array [
    "@/slices",
    Array [
      "CallToAction",
    ],
  ],
  Array [
    "vue-essential-slices",
    Array [
      "CallToAction",
    ],
  ],
]
`;
